File: 2807.insert-greatest-common-divisors-in-linked-list.java
#
# @lc app=leetcode id=2807 lang=java
#
# [2807] Insert Greatest Common Divisors in Linked List
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode curr = head;
        while (curr != null && curr.next != null) {
            int g = gcd(curr.val, curr.next.val);
            ListNode newNode = new ListNode(g, curr.next);
            curr.next = newNode;
            curr = newNode.next;
        }
        return head;
    }
    
    private static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}
# @lc code=end

File: 2807.insert-greatest-common-divisors-in-linked-list.go
#
# @lc app=leetcode id=2807 lang=golang
#
# [2807] Insert Greatest Common Divisors in Linked List
#
# @lc code=start
/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a % b
    }
    return a
}

func insertGreatestCommonDivisors(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    current := head
    for current != nil && current.Next != nil {
        // Calculate GCD of current and next node values
        gcdValue := gcd(current.Val, current.Next.Val)
        
        // Create new node with GCD value
        newNode := &ListNode{Val: gcdValue}
        
        // Insert new node between current and current.Next
        newNode.Next = current.Next
        current.Next = newNode
        
        // Move to the next original node (skip the inserted node)
        current = newNode.Next
    }
    
    return head
}
# @lc code=end